api/src/main/java/ai/djl/translate/SeqBatchScheduler.java

```
@@ -0,0 +1,144 @@
+package ai.djl.translate;
+
+import ai.djl.ndarray.NDArray;
+import ai.djl.ndarray.NDManager;
+import ai.djl.ndarray.index.NDIndex;
+import ai.djl.ndarray.types.DataType;
+import ai.djl.ndarray.types.Shape;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+// This is a scheduler, serving as an API to the consumer of the systme, allowing for three major
+// actions: initForward, addBatch, fastForward, collectResults.
+// An optimal control sequence should be solved, after considering the time consumption of each
+// action, the batch size and sequence length of queueing requests. Such optimal control solver
+// needs additional effort. Primitive policy is setting several thresholds.
+public abstract class SeqBatchScheduler {
+    LMBlock lmBlock;
+    SeqBatcher seqBatcher;
+
+    NDManager manager;
+
+    SearchConfig config;
+
+    Map<Long, NDArray> results;
+
+    public SeqBatchScheduler(LMBlock lmBlock) {
+        this.lmBlock = lmBlock;
+        results = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Initialize the iteration and SeqBatcher
+     *
+     * @return SeqBatcher. Stores the search state and operate on the BatchList.
+     */
+    public abstract SeqBatcher initForward(
+            NDArray inputIds, NDArray batchUids, SearchConfig config);
+
+    /** Go forward for a given number of iterations. */
+    public void incrementForward(int count) throws NullPointerException {
+        int i = 0;
+        while (i++ < count) {
+            if (seqBatcher == null || seqBatcher.getData() == null) {
+                System.out.println(
```

Should probably throw an Exception here